internal/stats/instrumentregistry/instrument_registry.go

```diff
@@ -0,0 +1,199 @@
+ *
+ */
+
+// Package instrumentregistry is a instrument registry that gRPC components can
+// use to register instruments (metrics).
+package instrumentregistry
+
+import (
+	"log"
+
+	"google.golang.org/grpc/experimental/stats/instrumentregistry"
+	"google.golang.org/grpc/stats"
+)
+
+// InstrumentDescriptor is a data of a registered instrument (metric).
+type InstrumentDescriptor struct {
+	// Name is the name of this metric.
+	Name string
+	// Description is the description of this metric.
+	Description string
+	// Unit is the unit of this metric.
+	Unit string
+	// Labels are the required label keys for this metric.
+	Labels []string
+	// OptionalLabels are the optional label keys for this
+	// metric.
+	OptionalLabels []string
+	// Default is whether this metric is on by default.
+	Default bool
+}
+
+// All of these globals are written to at initialization time only, and are read
+// only after initialization.
+
+// Int64CountInsts is information about registered int 64 count instruments in
+// order of registration.
+var Int64CountInsts []InstrumentDescriptor
+
+// Float64CountInsts is information about registered float 64 count instruments
+// in order of registration.
+var Float64CountInsts []InstrumentDescriptor
+
+// Int64HistoInsts is information about registered int 64 histo instruments in
+// order of registration.
+var Int64HistoInsts []InstrumentDescriptor
+
+// Float64HistoInsts is information about registered float 64 histo instruments
+// in order of registration.
+var Float64HistoInsts []InstrumentDescriptor
+
+// Int64GaugeInsts is information about registered int 64 gauge instruments in
+// order of registration.
+var Int64GaugeInsts []InstrumentDescriptor
+
+// registeredInsts are the registered instrument descriptor names.
+var registeredInsts = make(map[string]bool)
+
+// ClearInstrumentRegistryForTesting clears the instrument registry for testing
+// purposes only.
+func ClearInstrumentRegistryForTesting() {
+	Int64CountInsts = nil
+	Float64CountInsts = nil
+	Int64HistoInsts = nil
+	Float64HistoInsts = nil
+	Int64GaugeInsts = nil
+	registeredInsts = make(map[string]bool)
+	instrumentregistry.DefaultMetrics = make(map[stats.Metric]bool)
+}
+
+func registerInst(name string, def bool) {
+	if registeredInsts[name] {
+		log.Panicf("instrument %v already registered", name)
+	}
+	registeredInsts[name] = true
+	if def {
+		instrumentregistry.DefaultMetrics[stats.Metric(name)] = true
+	}
+}
+
+// RegisterInt64Count registers the int count instrument description onto the
+// global registry. It returns a typed handle to use when recording data.
+//
+// NOTE: this function must only be called during initialization time (i.e. in
+// an init() function), and is not thread-safe. If multiple instruments are
+// registered with the same name, this function will panic.
+func RegisterInt64Count(name string, desc string, unit string, labels []string, optionalLabels []string, def bool) instrumentregistry.Int64CountHandle {
```

Comment:
How about having the user pass the `InstrumentDescriptor` directly?

Reply:
Hmm ok. Thinking about it that would allow scaling up fields if needed in a backward compatible way if we decide we want to persist another piece of information here. Underlying though I'll persist around a copy of the struct not the same pointer as a user.

Reply:
You can pass by value instead of pointer to avoid the need to manually copy.

Reply:
I was thinking about it too now that I use a metrics set for defaults and there's a metrics type I might just have the inst desc take a Metric, and have the user pass in a Metric for explicit types and no need to typecast in this registry.